### PR TITLE
test: setup e2e test

### PR DESCRIPTION
**Pull request by anon-pradip**
Added e2e tests setup along with configurations.
Implemented test scenarios for previewing slides in presentation viewer (including uploading of markdown file using API).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,8 @@
   },
   "dependencies": {
     "@types/reveal.js": "^4.4.8",
+    "axios": "^1.6.7",
+    "dotenv": "^16.4.5",
```
Comment (saw-jan) [inline package.json:15]: these should be in devDependencies

File: tests/package.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (grgprarup) [inline tests/package.json:1]: You can use the base package.json file. There is no need to create a separate package.json for tests/.

File: tests/e2e/features/presentationViewer.feature [hunk 3]
```diff
@@ -0,0 +1,29 @@
+Feature: Markdown presentation viewer
+  As a user
+  I want to view the markdown in the presentation format
+  So that I can view the markdown documents in the presentation format
+  
+  Background:
+    Given the user has logged in with username "admin" and password "admin"
+    
+  
+  Scenario: Open Markdown file and verify rendering
```
Comment (grgprarup) [inline tests/e2e/features/presentationViewer.feature:10]: It's better to start with a single scenario.

File: tests/e2e/sampleMdFIle.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+## Slide 1
```
Comment (grgprarup) [inline tests/e2e/sampleMdFIle.md:1]: Add the test file inside dir `filesForUpload`

File: tests/e2e/features/presentationViewer.feature [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    Given the user opens a markdown file "index1.md"
+    When the file is rendered
+    Then the content should be displayed in a presentation viewer
+    And all tags should be rendered correctly (e.g., # -> <h1>)
+
+
+  Scenario: Verify slide content and navigation
+    Given the user opens a markdown file with multiple slides "slides.md"
```
Comment (grgprarup) [inline tests/e2e/features/presentationViewer.feature:18]: We need to create or upload the markdown file.

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+Given(
+  "the user has logged in with username {string} and password {string}",
+  async function (username, password) {
+    await page.goto(signInPageUrl);
+    const loginFormLocator = page.locator(loginForm);
+    await expect(loginFormLocator).toBeVisible(); 
+    await page.locator(homePage);
+    await page.fill(usernameInputField, username);
+    await page.fill(passwordInputField, password);
+    await page.click(loginButton);
```
Comment (grgprarup) [inline tests/e2e/stepDefinitions/presentationViewerContext.js:29]: Implement page object for this.

File: tests/e2e/features/presentationViewer.feature (hunk 5 above)
Comment (koebel) [reply tests/e2e/features/presentationViewer.feature:18]: I guess that would be part of background since it applies to all scenarios...

File: tests/e2e/features/presentationViewer.feature [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    Then the total number of slides should be correct
```
Comment (koebel) [inline tests/e2e/features/presentationViewer.feature:19]: Usually the step order is `Given`, `When`, `Then`. It seems like here the `When`-step is skipped...
What about moving this `Then`-statement to the end of this scenario?

File: tests/e2e/features/presentationViewer.feature [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    When the user navigates to the next slide
+    Then the content of the current slide should be displayed
+    And the slide number indicator should reflect the current slide
+  
+
+  Scenario: Navigate slides using navigation buttons
+    Given the user opens a markdown file with multiple slides "slides.md"
+    Then the presentation viewer should have navigation buttons (Previous, Next)
```
Comment (koebel) [inline tests/e2e/features/presentationViewer.feature:27]: I don't think this step is needed

File: tests/e2e/sampleMdFIle.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+A paragraph edited finally content one with some text and a [link](https://hakim.se).
```
Comment (koebel) [inline tests/e2e/sampleMdFIle.md:2]: any reason we use the link to this particular company URL in the test content? I think in general we would rather use a non-proprietary URL...

File: tests/e2e/features/presentationViewer.feature [hunk 10]
```diff
@@ -0,0 +1,15 @@
+Feature: Markdown presentation viewer
+  As a user
+  I want to view the markdown in the presentation format
+  So that I can view the markdown documents in the presentation format
```
Comment (koebel) [inline tests/e2e/features/presentationViewer.feature:4]: Please make a clear distinction between **what** the user wants to do (`I want to ...`) and **why** the user wants to do this (`so that ...`).  
It might be helpful to have a look at the user story that was initially defined for this extension app in [this issue](https://github.com/JankariTech/web-app-presentation-viewer/issues/1)

File: tests/package.json (hunk 2 above)
Comment (anon-pradip) [reply tests/package.json:1]: since common js is used in the tests/, separate package.json is used for that

File: tests/e2e/features/slideViewer.feature [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+  As a user
+  I want to view markdown documents in presentation format
+  So that I can easily present content in a structured manner
+  
+  Background:
+    Given user "Admin" has logged in
+
+
+  Scenario: preview markdown file in presentation viewer
+    Given user has uploaded the markdown file "test-markdown.md"
```
Comment (individual-it) [inline tests/e2e/features/slideViewer.feature:11]: when not mentioning the user name it should be `the user`
see also https://doc.owncloud.com/server/next/developer_manual/testing/acceptance-tests.html#specifying-the-actor

File: tests/e2e/features/slideViewer.feature [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    When user previews a markdown file "test-markdown.md" in presentation viewer mode
+    Then markdown content should be rendered as slides in a presentation format
```
Comment (individual-it) [inline tests/e2e/features/slideViewer.feature:13]:
```suggestion
    Then the markdown content should be rendered as slides in a presentation format
```

File: tests/e2e/pageObjects/slideViewerPage.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    await page.fill(this.passwordInputFieldSelector, password)
+    await page.click(this.loginBtnSelector)
+  }
+
+  async openMDFileInPresentationViewer() {
+    await page.click(this.contextMenuBtnSelector)
+    await page.click(this.openInPresentationViewerBtnSelector)
+  }
+
+  async clickNextSlideBtnEnabled() {
```
Comment (individual-it) [inline tests/e2e/pageObjects/slideViewerPage.js:36]: I'm not sure about the naming, what is the action this button does?

File: tests/e2e/utils/helpers.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    Authorization: `Basic ${Buffer.from(`${config.adminUser}:${config.adminPassword}`).toString(
+      'base64'
+    )}`,
+    'Content-Type': 'text/markdown'
+  }
+  return headers
+}
+
+const uploadFileToServer = async (fileName, method) => {
+  const fileUploadUrl = `${baseUrlOcis}/dav/files/admin/${fileName}`
```
[the remark this answers is not in the document]
Comment (individual-it) [reply tests/e2e/utils/helpers.js:38]: same applies to 
- getAuthHeaders
- deleteFiles

File: tests/e2e/utils/helpers.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  const headers = {
+    Authorization: `Basic ${Buffer.from(`${config.adminUser}:${config.adminPassword}`).toString(
+      'base64'
+    )}`,
+    'Content-Type': 'text/markdown'
+  }
+  return headers
+}
+
+const uploadFileToServer = async (fileName, method) => {
```
Comment (individual-it) [inline tests/e2e/utils/helpers.js:37]: why is the `method` parameter needed here? an upload always uses `PUT`

File: tests/e2e/utils/helpers.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  const fileUploadUrl = `${baseUrlOcis}/dav/files/admin/${fileName}`
+  const filePath = path.join(config.assets, fileName)
+  const fileContent = fs.readFileSync(filePath, 'utf8')
+  await makeApiRequest(method, fileUploadUrl, getAuthHeaders(), fileContent)
+  createdFiles.push(fileName)
+}
+
+const createFile = async (fileName) => {
```
Comment (individual-it) [inline tests/e2e/utils/helpers.js:45]: why is this function needed? Its just a wrapper for a single line function

File: tests/e2e/stepDefinitions/slideViewerContext.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+When(
+  'user previews a markdown file {string} in presentation viewer mode',
+  async function (fileName) {
+    await slideViewerPage.openMDFileInPresentationViewer()
+    await expect(page.locator(slideViewerPage.slideViewerHomepageSelector)).toBeVisible()
+  }
+)
+
+Then('markdown content should be rendered as slides in a presentation format', async function () {
+  await expect(page.locator(slideViewerPage.slidesContainerSelector)).toBeVisible()
```
Comment (individual-it) [inline tests/e2e/stepDefinitions/slideViewerContext.js:28]: that is a very basic check, but I think that is fine for now and more specific checks can be added later

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/e2e/pageObjects/slideViewerPage.js (hunk 13 above)
Comment (anon-pradip) [reply tests/e2e/pageObjects/slideViewerPage.js:36]: This function just checks if the file has multiple sides or not. If nextSlideBtnEnabledSelector is found then the file has multiple slides otherwise not. Since this function is not being used in the current scenario, I am removing it.

Comment (anon-pradip) [reply tests/e2e/pageObjects/slideViewerPage.js:36]: This button selects the option "Open in Presentation Viewer" from the menu containing various options like Open in Text Editor, Open in Presentation Viewer, Share, Copy link, Download and so on.

Comment (anon-pradip) [reply tests/e2e/pageObjects/slideViewerPage.js]: This function generally checks if the "next slide" button exists or not. Since it is not used in this scenario, I removed this function for now.

File: tests/e2e/utils/helpers.js (hunk 15 above)
Comment (anon-pradip) [reply tests/e2e/utils/helpers.js]: since `makeApiRequest()` is dynamic and takes `method` as parameter, the function `uploadFileToServer()` is also made to take `method` as parameter. But passed `PUT` directly as parameter in the `makeApiRequest` inside `uploadFileToServer` now.

File: .gitignore [hunk 18]
```diff
@@ -6,3 +6,6 @@ dist
 .idea
 
 /.pnpm-store/
+.env
+
+tests/pnpm-lock.yaml
```
Comment (saw-jan) [inline .gitignore:11]: we want to keep the lock file if any packages are added to test level in the future
```suggestion
```

File: package.json [hunk 19]
```diff
@@ -9,22 +9,28 @@
...
     "@ownclouders/prettier-config": "0.0.1",
+    "@playwright/test": "^1.42.1",
     "@typescript-eslint/eslint-plugin": "^7.1.1",
+    "axios": "^1.6.7",
     "eslint": "^8.57.0",
     "eslint-config-prettier": "^9.1.0",
     "eslint-plugin-prettier-vue": "^5.0.0",
     "eslint-plugin-unused-imports": "^3.1.0",
     "eslint-plugin-vue": "^9.22.0",
+    "ts-node": "^10.9.2",
```
Comment (saw-jan) [inline package.json:33]: playwright tests are in js so this should not be required I think
```suggestion
```

File: tests/e2e/cucumber.conf.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
+const config = {
+  // environment
+  baseUrlOcis: process.env.BASE_URL_OCIS ?? 'https://host.docker.internal:9200',
```
Comment (saw-jan) [inline tests/e2e/cucumber.conf.js:3]:
```suggestion
  baseUrlOcis: process.env.BASE_URL_OCIS || 'https://host.docker.internal:9200',
```

Comment (saw-jan) [inline tests/e2e/cucumber.conf.js:3]: Can it be renamed to ...
```suggestion
  baseUrlOcis: process.env.OCIS_URL ?? 'https://host.docker.internal:9200',
```

File: tests/e2e/cucumber.conf.js [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+  assets: './tests/e2e/filesForUpload',
+  adminUser: process.env.ADMIN_USERNAME || 'admin',
+  adminPassword: process.env.ADMIN_PASSWORD || 'admin',
+  normalUserUsername: process.env.NORMAL_USER_USERNAME || 'marie',
+  normalUserPassword: process.env.NORMAL_USER_PASSWORD || 'marie',
```
Comment (saw-jan) [inline tests/e2e/cucumber.conf.js:8]: I would not keep normal user/pass here
```suggestion
```

File: tests/e2e/cucumber.hooks.js [hunk 22]
```diff
@@ -0,0 +1,31 @@
+const { Before, BeforeAll, AfterAll, After, setDefaultTimeout } = require('@cucumber/cucumber')
+const { firefox } = require('playwright')
+const { deleteFiles } = require('./utils/helpers')
+// default timeout (in milliseconds)
+setDefaultTimeout(100000)
```
Comment (saw-jan) [inline tests/e2e/cucumber.hooks.js:5]: 60 second timeout should be enough or was it not?
```suggestion
setDefaultTimeout(60 * 1000)
```

File: tests/e2e/cucumber.hooks.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+// launch the browser
+BeforeAll(async function () {
+  global.browser = await firefox.launch({
```
Comment (saw-jan) [inline tests/e2e/cucumber.hooks.js:9]: Let's use chrome browser because it has the largest user base
```suggestion
  global.browser = await chrome.launch({
```

File: tests/e2e/cucumber.hooks.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    headless: false,
+    slowMo: 0
```
Comment (saw-jan) [inline tests/e2e/cucumber.hooks.js:11]: the values should be taken from config
```suggestion
    headless: <from-config>,
    slowMo: <from-config>
```

File: cucumber.cjs [hunk 25]
```diff
@@ -0,0 +1,9 @@
+const {config} = require('./tests/e2e/cucumber.conf.js')
```
Comment (saw-jan) [inline cucumber.cjs:1]: Please, rebase the PR and run `pnpm lint`

File: tests/e2e/features/slideViewer.feature [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+  So that I can easily present content in a structured manner
+  
+  Background:
+    Given user "admin" has logged in
+
+
+  Scenario: preview markdown file in presentation viewer
+    Given user "admin" has uploaded the markdown file "test-markdown.md"
+
+    When user "admin" previews a markdown file "test-markdown.md" in presentation viewer mode
```
Comment (saw-jan) [inline tests/e2e/features/slideViewer.feature:13]:
```suggestion
    When user "admin" previews a markdown file "test-markdown.md" in the presentation viewer
```

File: tests/e2e/pageObjects/ocisPage.js [hunk 27]
```diff
@@ -0,0 +1,23 @@
+const { config } = require('../cucumber.conf')
+
+class OcisPage {
```
Comment (saw-jan) [inline tests/e2e/pageObjects/ocisPage.js:3]: Also use the same naming style for the class file: `ocisPage.js` -> `OcisPage.js`

File: tests/e2e/pageObjects/slideViewerPage.js [hunk 28]
```diff
@@ -0,0 +1,16 @@
+class SlideViewerPage {
```
Comment (saw-jan) [inline tests/e2e/pageObjects/slideViewerPage.js:1]:
```suggestion
class PresentationViewerPage {
```

File: tests/e2e/pageObjects/slideViewerPage.js [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+  constructor() {
+    this.filesContainerSelector = '#files-view'
+    this.contextMenuBtnSelector = '.resource-table-btn-action-dropdown'
+    this.openInPresentationViewerBtnSelector = '.oc-files-actions-presentation-viewer-trigger'
+    this.slidesContainerSelector = '#presentation-viewer-main .slides'
+    this.slideViewerHomepageSelector = '#presentation-viewer-main'
+  }
+
+  async openMDFileInPresentationViewer() {
```
Comment (saw-jan) [inline tests/e2e/pageObjects/slideViewerPage.js:10]: I believe this action can go to `OcisPage`

File: tests/e2e/stepDefinitions/slideViewerContext.js [hunk 30]
```diff
@@ -0,0 +1,34 @@
+const { Given, When, Then } = require('@cucumber/cucumber')
+const { expect } = require('@playwright/test')
+const { SlideViewerPage } = require('../pageObjects/slideViewerPage')
+const { uploadFileToServer, getUserCredentials } = require('../utils/helpers')
+const { OcisPage } = require('../pageObjects/ocisPage')
+const slideViewerPage = new SlideViewerPage()
+const ocisPage = new OcisPage()
```
Comment (saw-jan) [inline tests/e2e/stepDefinitions/slideViewerContext.js:7]:
```suggestion
const presentationViewer = new SlideViewerPage()
const ocis = new OcisPage()
```

File: tests/e2e/utils/helpers.js [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+  }
+}
+
+const getUserCredentials = (user) => {
+  if (user === 'admin') {
+    return { username: config.adminUser, password: config.adminPassword }
+  } else {
+    throw new Error('user not found!')
+  }
+}
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:35]: this should go to another separate file

File: tests/e2e/utils/helpers.js [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+
+const getAuthHeaders = (user) => {
+  try {
+    const { username, password } = getUserCredentials(user)
+    const headers = {
+      Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`,
+      'Content-Type': 'text/markdown'
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:42]: only return auth header
```suggestion
```

File: tests/e2e/utils/helpers.js [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+    }
+    return headers
+  } catch (error) {
+    throw new Error(error.message)
+  }
+}
+
+const uploadFileToServer = async (fileName, user) => {
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:50]:
```suggestion
const uploadFile = async (fileName, user) => {
```

File: tests/e2e/utils/helpers.js [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+  const fileUploadUrl = `${baseUrlOcis}/dav/files/${user}/${fileName}`
+  const filePath = path.join(config.assets, fileName)
+  const fileContent = fs.readFileSync(filePath, 'utf8')
+  await makeApiRequest('PUT', fileUploadUrl, getAuthHeaders(user), fileContent)
+  userDetails.push({ user, fileName })
+}
+
+const deleteFiles = async () => {
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:58]:
```suggestion
const deleteFile = async () => {
```

File: tests/e2e/utils/helpers.js [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+      'Content-Type': 'text/markdown'
+    }
+    return headers
+  } catch (error) {
+    throw new Error(error.message)
+  }
+}
+
+const uploadFileToServer = async (fileName, user) => {
+  const fileUploadUrl = `${baseUrlOcis}/dav/files/${user}/${fileName}`
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:51]: use function to build urls
```suggestion
  const fileUploadUrl = getWebDavUrl(user, fileName)
```

File: tests/e2e/cucumber.hooks.js (hunk 23 above)
Comment (anon-pradip) [reply tests/e2e/cucumber.hooks.js:9]: [Chrome and ms-edge issue with playwright](https://playwright.dev/docs/browsers#google-chrome--microsoft-edge). Also chrome is not available as direct installation in playwright. Shall we use chromium?

Comment (saw-jan) [reply tests/e2e/cucumber.hooks.js:9]: yeah please chromium

File: tests/e2e/utils/helpers.js (hunk 32 above)
Comment (anon-pradip) [reply tests/e2e/utils/helpers.js:42]: Since we are playing with the markdown file, won't it be better to put `'Content-Type': 'text/markdown'`?

File: cucumber.cjs [hunk 36]
```diff
@@ -0,0 +1,9 @@
+const { config } = require('./tests/e2e/cucumber.conf.js')
```
Comment (saw-jan) [inline cucumber.cjs:1]:
```suggestion
```

File: cucumber.cjs [hunk 37]
```diff
@@ -0,0 +1,9 @@
...
+
+module.exports = {
+  default: {
+    require: ['tests/e2e/**/*.js'],
+    retry: config.retry,
```
Comment (saw-jan) [inline cucumber.cjs:6]:
```suggestion
    retry: process.env.RETRY || 0,
```

File: tests/e2e/cucumber.conf.js [hunk 38]
```diff
@@ -0,0 +1,16 @@
+const config = {
+  // environment
+  baseOcisUrl: process.env.OCIS_URL || 'https://host.docker.internal:9200',
+  assets: './tests/e2e/filesForUpload',
+  adminUser: process.env.ADMIN_USERNAME || 'admin',
+  adminPassword: process.env.ADMIN_PASSWORD || 'admin',
+  // cucumber
+  retry: process.env.RETRY || 0,
```
Comment (saw-jan) [inline tests/e2e/cucumber.conf.js:8]:
```suggestion
```

**Review by saw-jan — CHANGES_REQUESTED** [review]
use env directly in cucumber config file

File: tests/e2e/cucumber.hooks.js [hunk 39]
```diff
@@ -0,0 +1,32 @@
+const { Before, BeforeAll, AfterAll, After, setDefaultTimeout } = require('@cucumber/cucumber')
```
Comment (saw-jan) [inline tests/e2e/cucumber.hooks.js:1]: Please, rename the file to `hooks.ts`

File: tests/e2e/features/presentationViewer.feature [hunk 40]
```diff
@@ -0,0 +1,13 @@
+Feature: markdown presentation viewer
+  As a user
+  I want to view markdown documents in presentation format
+  So that I can easily present content in a structured manner
+  
+  Background:
+    Given user "admin" has logged in
```
[the remark this answers is not in the document]
Comment (koebel) [reply tests/e2e/features/presentationViewer.feature:7]: the order of these two steps doesn't seem logical for me. I guess the user "admin" first needs to log in before being able to upload a file...

[the remark this answers is not in the document]
Comment (koebel) [reply tests/e2e/features/presentationViewer.feature:7]: that might be a little clearer. Anyway, I guess you can also imagine that user "admin" had uploaded the files sometimes in the past (it doesn't matter if he/she was using the API or the UI) and then logs in again and goes through the other steps outlined in the scenario... In this case your order would also be correct.

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 41]
```diff
@@ -0,0 +1,36 @@
...
+Given('user {string} has logged in', async function (user) {
+  await ocis.navigateToLoginPage()
+  const { username, password } = getUserCredentials(user)
+  await ocis.login(username, password)
+  await expect(page.locator(ocis.filesContainerSelector)).toBeVisible()
+})
+
+Given('user {string} has uploaded the markdown file {string}', async function (user, fileName) {
+  await uploadFile(fileName, user)
+  await page.reload()
```
[the remark this answers is not in the document]
Comment (anon-pradip) [reply tests/e2e/stepDefinitions/presentationViewerContext.js:20]: page should be reloaded to see the uploaded the files since we are uploading through API and the file wont be shown without reloading the paage

File: docker-compose.yml [hunk 42]
```diff
@@ -7,7 +7,7 @@ services:
       # OCIS
       OCIS_URL: https://host.docker.internal:9200
       OCIS_INSECURE: true
-      OCIS_LOG_LEVEL: error
+
```
Comment (saw-jan) [inline docker-compose.yml:10]: the suggestion was only to undo the unnecessary change. Please, add the log level
```suggestion
      OCIS_LOG_LEVEL: error
```

Comment (saw-jan) [reply docker-compose.yml:10]: This file should not be touched in this PR

File: tests/e2e/hooks.js [hunk 43]
```diff
@@ -0,0 +1,32 @@
+const { Before, BeforeAll, AfterAll, After, setDefaultTimeout } = require('@cucumber/cucumber')
+const { deleteFile } = require('./utils/helpers')
+const config = require('./config')
+const { chromium } = require('@playwright/test')
+// default timeout (in milliseconds)
```
Comment (saw-jan) [inline tests/e2e/hooks.js:5]: good to have separation between imorts and others
```suggestion
const { chromium } = require('@playwright/test')

// default timeout (in milliseconds)
```

File: tests/e2e/utils/helpers.js [hunk 44]
```diff
@@ -0,0 +1,64 @@
+const axios = require('axios')
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:1]: rename `helpers.js` -> `apiHelper.js`

File: tests/e2e/utils/helpers.js [hunk 45]
```diff
@@ -0,0 +1,64 @@
...
+  await makeApiRequest('PUT', fileUploadUrl, getAuthHeaders(user), fileContent)
+  userDetails.push({ user, fileName })
+}
+
+const deleteFile = async () => {
+  for (const userDetail of userDetails) {
+    const fileDeleteUrl = getWebDavUrl(userDetail.user, userDetail.fileName)
+    await makeApiRequest('DELETE', fileDeleteUrl, getAuthHeaders(userDetail.user), null)
+  }
+}
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:57]:
```suggestion
const deleteResource = (user, resource) => {
    const url = getWebDavUrl(user, resource)
    return makeApiRequest('DELETE', url, getAuthHeaders(user))
}
const cleanupResources = async () => {
  for (const userDetail of userDetails) {
    await deleteResource(userDetail.user, userDetail.fileName)
  }
}
```

File: tests/e2e/utils/helpers.js [hunk 46]
```diff
@@ -0,0 +1,64 @@
...
+  return `${baseOcisUrl}/dav/files/${user}/${fileName}`
+}
+
+const uploadFile = async (fileName, user) => {
+  const fileUploadUrl = getWebDavUrl(user, fileName)
+  const filePath = path.join(config.assets, fileName)
+  const fileContent = fs.readFileSync(filePath, 'utf8')
+  await makeApiRequest('PUT', fileUploadUrl, getAuthHeaders(user), fileContent)
+  userDetails.push({ user, fileName })
+}
```
Comment (saw-jan) [inline tests/e2e/utils/helpers.js:50]: pass only `user`, `filename` and `content`. read the content else where 
```suggestion
const uploadFile = (user, filename, content="") => {
  const url = getWebDavUrl(user, filename)
  return makeApiRequest('PUT', url, getAuthHeaders(user), content)
}
```

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 47]
```diff
@@ -0,0 +1,38 @@
...
+const { getUserCredentials } = require('../utils/userHelper')
+const Ocis = require('../pageObjects/OcisPage')
+
+const presentationViewer = new PresentationViewer()
+const ocis = new Ocis()
+
+Given(
+  'user {string} has uploaded the markdown file {string} using API',
+  async function (user, fileName) {
+    await uploadFile(fileName, user)
```
Comment (saw-jan) [inline tests/e2e/stepDefinitions/presentationViewerContext.js:14]: need a check to make sure that upload was successful

File: tests/e2e/utils/helpers.js (hunk 44 above)
Comment (anon-pradip) [reply tests/e2e/utils/helpers.js:1]: I think it would be better to place `uploadFile` and `deleteFile` to a new file `fileHelper.js` and keep the remaining code in the same file renamed to `apiHelper.js`, isn't it?

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 48]
```diff
@@ -0,0 +1,40 @@
...
+const presentationViewer = new PresentationViewer()
+const ocis = new Ocis()
+
+Given(
+  'user {string} has uploaded the markdown file {string} using API',
+  async function (user, fileName) {
+    const response = await uploadFile(fileName, user)
+    assert.equal(response.status, 201, 'Failed to upload file!')
+  }
+)
```
Comment (anon-pradip) [inline tests/e2e/stepDefinitions/presentationViewerContext.js:18]: @saw-jan Did you mean this way of providing meaningful message while uploading file?

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 49]
```diff
@@ -0,0 +1,40 @@
+const { Given, When, Then } = require('@cucumber/cucumber')
+const { expect } = require('@playwright/test')
+const PresentationViewer = require('../pageObjects/PresentationViewerPage')
+const assert = require('node:assert')
```
Comment (saw-jan) [inline tests/e2e/stepDefinitions/presentationViewerContext.js:4]:
```suggestion
const { expect } = require('@playwright/test')
const assert = require('assert')
const PresentationViewer = require('../pageObjects/PresentationViewerPage')
```

File: tests/e2e/features/presentationViewer.feature [hunk 50]
```diff
@@ -0,0 +1,13 @@
...
+  So that I can easily present content in a structured manner
+  
+  Background:
+    Given user "admin" has uploaded the markdown file "test-markdown.md" using API
+    And user "admin" has logged in
+
+
+  Scenario: preview markdown file in presentation viewer
+    When user "admin" previews a markdown file "test-markdown.md" in presentation viewer
+    Then markdown file "test-markdown.md" should be opened and rendered in the presentation viewer
```
Comment (saw-jan) [inline tests/e2e/features/presentationViewer.feature:13]:
```suggestion
    Then markdown file "test-markdown.md" should be opened in the presentation viewer
```

File: tests/e2e/utils/apiHelper.js [hunk 51]
```diff
@@ -0,0 +1,46 @@
+const axios = require('axios')
+const https = require('https')
+const config = require('../config')
+const { getUserCredentials } = require('./userHelper')
+
+const baseOcisUrl = config.baseOcisUrl
```
Comment (saw-jan) [inline tests/e2e/utils/apiHelper.js:6]:
```suggestion
```

File: tests/e2e/utils/apiHelper.js [hunk 52]
```diff
@@ -0,0 +1,46 @@
...
+      Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`
+    }
+    return headers
+  } catch (error) {
+    throw new Error(error.message)
+  }
+}
+
+const getWebDavUrl = (user, fileName) => {
+  return `${baseOcisUrl}/dav/files/${user}/${fileName}`
```
Comment (saw-jan) [inline tests/e2e/utils/apiHelper.js:39]:
```suggestion
  return `${config.baseOcisUrl}/dav/files/${user}/${fileName}`
```

File: tests/e2e/utils/apiHelper.js [hunk 53]
```diff
@@ -0,0 +1,46 @@
...
+    const response = await axios({
+      method,
+      url,
+      headers,
+      data,
+      httpsAgent: new https.Agent({ rejectUnauthorized: false })
+    })
+    return response
+  } catch (error) {
+    const errorMessage = `API request failed: ${error.message} - ${url} - ${method} - Headers: ${JSON.stringify(
```
Comment (saw-jan) [inline tests/e2e/utils/apiHelper.js:19]:
```suggestion
    const errorMessage = `API request failed: ${error.message}\nURL: ${url}\nMETHOD: ${method}\nHEADERS: ${JSON.stringify(
```

File: tests/e2e/utils/fileHelper.js [hunk 54]
```diff
@@ -0,0 +1,34 @@
...
+const path = require('path')
+const { getWebDavUrl, makeApiRequest, getAuthHeaders } = require('./apiHelper')
+const config = require('../config')
+
+const userDetails = []
+
+const getFileContent = (fileName) => {
+  const filePath = path.join(config.assets, fileName)
+  const fileContent = fs.readFileSync(filePath, 'utf8')
+  return fileContent
```
Comment (saw-jan) [inline tests/e2e/utils/fileHelper.js:11]:
```suggestion
  return fs.readFileSync(filePath, 'utf8')
```

File: tests/e2e/utils/fileHelper.js [hunk 55]
```diff
@@ -0,0 +1,34 @@
...
+  const content = getFileContent(fileName)
+  const response = await makeApiRequest('PUT', url, getAuthHeaders(user), content)
+  userDetails.push({ user, fileName })
+  return response
+}
+
+const deleteResource = (user, resource) => {
+  const url = getWebDavUrl(user, resource)
+  const response = makeApiRequest('DELETE', url, getAuthHeaders(user))
+  return response
```
Comment (saw-jan) [inline tests/e2e/utils/fileHelper.js:25]:
```suggestion
  retrun makeApiRequest('DELETE', url, getAuthHeaders(user))
```

File: tests/e2e/utils/fileHelper.js [hunk 56]
```diff
@@ -0,0 +1,34 @@
+const fs = require('fs')
+const path = require('path')
+const { getWebDavUrl, makeApiRequest, getAuthHeaders } = require('./apiHelper')
+const config = require('../config')
+
+const userDetails = []
+
+const getFileContent = (fileName) => {
```
Comment (saw-jan) [inline tests/e2e/utils/fileHelper.js:8]:
```suggestion
const getLocalFileContent = (fileName) => {
```

File: tests/e2e/stepDefinitions/presentationViewerContext.js [hunk 57]
```diff
@@ -0,0 +1,40 @@
...
+const { uploadFile } = require('../utils/fileHelper')
+
+const presentationViewer = new PresentationViewer()
+const ocis = new Ocis()
+
+Given(
+  'user {string} has uploaded the markdown file {string} using API',
+  async function (user, fileName) {
+    const response = await uploadFile(fileName, user)
+    assert.equal(response.status, 201, 'Failed to upload file!')
```
Comment (saw-jan) [inline tests/e2e/stepDefinitions/presentationViewerContext.js:16]: assertion is not required here because `makeApiRequest` already handles the request error
```suggestion
    await uploadFile(fileName, user)
```

File: tests/e2e/utils/markdown.js [hunk 58]
```diff
@@ -0,0 +1,36 @@
+const markdown = require('markdown-it')
+const { getLocalFileContent } = require('./fileHelper')
+
+const md = markdown()
+
+const sanitizeRenderedContent = (content) => {
+  return content
+    .replace(/&quot;/g, '"')
+    .replace(/\n(?=<\/code><\/pre>)/g, '')
```
Comment (saw-jan) [inline tests/e2e/utils/markdown.js:9]: hmm?

File: tests/e2e/utils/apiHelper.js [hunk 59]
```diff
@@ -0,0 +1,44 @@
...
+  try {
+    const { username, password } = getUserCredentials(user)
+    const headers = {
+      Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`
+    }
+    return headers
+  } catch (error) {
+    throw new Error(error.message)
+  }
+}
```
Comment (saw-jan) [inline tests/e2e/utils/apiHelper.js:34]:
```suggestion
    const { username, password } = getUserCredentials(user)
    const headers = {
      Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`
    }
    return headers
}
```

File: tests/e2e/utils/markdown.js (hunk 58 above)
Comment (anon-pradip) [reply tests/e2e/utils/markdown.js]: Sorry. This file is from another branch  but pushed accidentally. Removed now

**Review by saw-jan — APPROVED** [review]
:+1: :sparkles:

**Review by individual-it — APPROVED** [review]
